### ISSUE TITLE
Identify hybrid add-ons

### ISSUE THREAD

**Issue by wagnerand** [mozilla/amo-validator#488]
Following: https://github.com/mozilla/addons-server/issues/2875

The validator should raise an *information* (displayed to reviewers only) with the following details:

Title: This add-on is a hybrid add-on.
Description: The folder `webextension` contains an embedded webextension. Handle validation messages with care. For more information, refer to https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Embedded_WebExtensions

**Comment by wagnerand** [comment]
I am wondering whether including the `webextension` folder in the scan does more good or more harm.

@kewisch @jvillalobos What do you think?

**Comment by rpl** [comment]
@wagnerand I just tried to run the existent amo-validator rules on the xpi files generated from the two example add-ons of the mdn/webextensions-examples repo:

- hybrid bootstrapped add-on: https://github.com/mdn/webextensions-examples/tree/master/embedded-webextension-bootstrapped/step1-hybrid-addon
- hybrid sdk add-on: https://github.com/mdn/webextensions-examples/tree/master/embedded-webextension-sdk/step1-hybrid-addon

By running the xpi files created from the above two add-on directories we can immediatelly notice the first change that we have to apply to the amo-validator:

```
Summary:
------------------------------
Detected type: Extension/Multi-Extension
------------------------------
Test failed! Errors:

....
Notice: Unrecognized element in install.rdf
```

which should instead detect the new `hasEmbeddedWebExtension` and show the notice message that @wagnerand suggests in this issue summary.

it is the first time that I look into the amo-validator sources, but it seems that we can achieve this by applying something similar to the following changes:

```diff
diff --git a/validator/testcases/installrdf.py b/validator/testcases/installrdf.py
index bdc71f7..70ca997 100644
--- a/validator/testcases/installrdf.py
+++ b/validator/testcases/installrdf.py
@@ -44,6 +44,7 @@ def _test_rdf(err, install):
                       'homepageURL',
                       'creator',
                       'multiprocessCompatible',
+                      'hasEmbeddedWebExtension',
                       'optionsType',
                       'type',
                       'updateKey',
@@ -93,6 +94,21 @@ def _test_rdf(err, install):
         elif predicate == 'multiprocessCompatible':
             err.save_resource('is_multiprocess_compatible', value == 'true')

+        elif predicate == 'hasEmbeddedWebExtension':
+            err.notice(('testcases_installrdf',
+                        '_test_rdf',
+                        'hasEmbeddedWebExtension'),
+                       'This add-on is an hybrid add-on',
+                       "The element \"%s\" was found in the add-on's "
+                       'install.rdf file.'
+                       'The folder "webextension" contains an embedded webextension.'
+                       'Handle validation messages with care. '
+                       'For more information, refer to '
+                       'https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Embedded_WebExtensions'
+                       % predicate,
+                       'install.rdf')
+            err.save_resource('has_embedded_webextension', value == 'true')
+
         # We skip over install.rdf tests during bulk validation. See bug 735841
         elif predicate in shouldnt_exist and not is_compat:
             err.error(('testcases_installrdf',
```

**Comment by mstriemer** [comment]
How is this pulled into addons-server? Will we need an issue there too?

**Comment by rpl** [comment]
@mstriemer I don't have a first hand knowledge of this part, but my guess is that the AMO server integrates the JSON version of the amo-validator results (from an amo-validator running in a docker container).

**Comment by EnTeQuAk** [comment]
@mstriemer this is being pushed in the `resources` key in generated JSON afair which you can work with in the validation results in addons-server

**Comment by jvillalobos** [comment]
> I am wondering whether including the webextension folder in the scan does more good or more harm.

I would expect the validations to have more in common than not, and skipping the folder entirely seems like a bigger risk. Wouldn't the legacy add-on code also have access to that folder, potentially doing shady things?

**Comment by wagnerand** [comment]
One thing that immediately comes into mind are iframes and the `type` attribute, which does not apply to webextensions. I'm sure there are others...

> Wouldn't the legacy add-on code also have access to that folder, potentially doing shady things?

Afaik, only through the messaging functions: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Embedded_WebExtensions#Exchanging_messages

**Comment by rpl** [comment]
>> Wouldn't the legacy add-on code also have access to that folder, potentially doing shady things?

> Afaik, only through the messaging functions: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Embedded_WebExtensions#Exchanging_messages

@jvillalobos is right, unfortunately, even if we are providing the messages exchange (which is safer) as the official API that let the two parts to collaborate, given that legacy add-on code is not really sandboxed, we can't totally prevent the legacy part to still trying to do something shady.
(but at least the opposite is true, the embedded webextension shouldn't be able to access the resources outside of the `webextension` dir, that is one of the reasons why we forced the embedded webextension to be always restricted to a subdir of the hybrid add-on).

I agree with Jorge: totally skipping the validation on the subdirectory sounds pretty risky to me.

**Comment by jvillalobos** [comment]
I favor running the amo-validator through the whole thing. The warning about hybrid extensions should give reviewers a hint about taking flags in the `webextension` folder with a grain of salt.

**Comment by wagnerand** [comment]
> Surfacing a notice to reviewers should at least allow them to extract the webextension folder and run addons-linter locally or as a standalone validation if they want.

Even admins do this very rarely. I don't expect any normal reviewer to even have the standalone validator installed. Downloading the XPI file, extracting the webextension part and submitting that to the standalone linter on AMO is too much of an effort for it to be reasonable.

**Comment by wagnerand** [comment]
I think this is done.
